File: Ejercicios_POO3.1_Java/src/Ejercicio1_Biblioteca/Biblioteca.java
package Ejercicio1_Biblioteca;


import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Biblioteca {
    private List<Libro> libros = new ArrayList<>();
    private List<Usuario> usuarios = new ArrayList<>();
    private Map<Usuario, List<Libro>> prestamos = new HashMap<>();

    public void agregarLibro(Libro libro) {
        libros.add(libro);
        System.out.println("Libro agregado: " + libro.getTitulo());
    }

    public void registrarUsuario(Usuario usuario) {
        usuarios.add(usuario);
        System.out.println("Usuario registrado: " + usuario.getNombre());
    }

    public void prestarLibro(Usuario usuario, String tituloLibro) {
        for (Libro libro : libros) {
            if (libro.getTitulo().equalsIgnoreCase(tituloLibro)) {
                if (libro.isPrestado()) {
                    System.out.println("El libro '" + tituloLibro + "' ya está prestado.");
                } else {
                    libro.prestar();
                    prestamos.computeIfAbsent(usuario, k -> new ArrayList<>()).add(libro);
                    System.out.println("Libro '" + tituloLibro + "' prestado a " + usuario.getNombre());
                }
                return;
            }
        }
        System.out.println("El libro '" + tituloLibro + "' no está disponible en la biblioteca.");
    }

    public void devolverLibro(Usuario usuario, String tituloLibro) {
        List<Libro> librosPrestados = prestamos.get(usuario);
        if (librosPrestados != null) {
            for (Libro libro : librosPrestados) {
                if (libro.getTitulo().equalsIgnoreCase(tituloLibro)) {
                    libro.devolver();
                    librosPrestados.remove(libro);
                    System.out.println("Libro '" + tituloLibro + "' devuelto por " + usuario.getNombre());
                    return;
                }
            }
        }
        System.out.println("El usuario " + usuario.getNombre() + " no tiene prestado el libro '" + tituloLibro + "'.");
    }

    public Usuario buscarUsuarioPorId(String id) {
        for (Usuario usuario : usuarios) {
            if (usuario.getId().equals(id)) {
                return usuario;
            }
        }
        System.out.println("Usuario no encontrado.");
        return null;
    }
}

File: Ejercicios_POO3.1_Kotlin/src/Ejercicio1_Biblioteca/Biblioteca.kt
package Ejercicio1_Biblioteca

class Biblioteca {
    private val libros = mutableListOf<Libro>()
    private val usuarios = mutableListOf<Usuario>()
    private val prestamos = mutableMapOf<Usuario, MutableList<Libro>>()

    fun agregarLibro(libro: Libro) {
        libros.add(libro)
        println("Libro agregado: ${libro.titulo}")
    }

    fun registrarUsuario(usuario: Usuario) {
        usuarios.add(usuario)
        println("Usuario registrado: ${usuario.nombre}")
    }

    fun prestarLibro(usuario: Usuario, tituloLibro: String) {
        val libro = libros.find { it.titulo.equals(tituloLibro, ignoreCase = true) }
        if (libro != null) {
            if (libro.prestado) {
                println("El libro '$tituloLibro' ya está prestado.")
            } else {
                libro.prestar()
                prestamos.getOrPut(usuario) { mutableListOf() }.add(libro)
                println("Libro '$tituloLibro' prestado a ${usuario.nombre}")
            }
        } else {
            println("El libro '$tituloLibro' no está disponible en la biblioteca.")
        }
    }

    fun devolverLibro(usuario: Usuario, tituloLibro: String) {
        val librosPrestados = prestamos[usuario]
        if (librosPrestados != null) {
            val libro = librosPrestados.find { it.titulo.equals(tituloLibro, ignoreCase = true) }
            if (libro != null) {
                libro.devolver()
                librosPrestados.remove(libro)
                println("Libro '$tituloLibro' devuelto por ${usuario.nombre}")
                return
            }
        }
        println("El usuario ${usuario.nombre} no tiene prestado el libro '$tituloLibro'.")
    }

    fun buscarUsuarioPorId(id: String): Usuario? {
        return usuarios.find { it.id == id } ?: run {
            println("Usuario no encontrado.")
            null
        }
    }
}